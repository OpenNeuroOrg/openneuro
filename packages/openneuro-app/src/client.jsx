/**
 * Browser client entrypoint - see server.tsx for SSR entrypoint
 */
import './scripts/utils/global-polyfill'
import { setupApm } from './scripts/apm.js'
import * as Sentry from '@sentry/browser'
import { ApolloProvider, InMemoryCache } from '@apollo/client'
import { createClient } from '@openneuro/client'
import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter, Route } from 'react-router-dom'
import App from './scripts/app'
import Index from './scripts/index'
import analyticsWrapper from './scripts/utils/analytics'
import { version } from './lerna.json'
import { config } from './scripts/config'
import * as gtag from './scripts/utils/gtag'

if (
  config.sentry.environment === 'production' ||
  config.sentry.environment === 'staging'
) {
  setupApm({
    serverUrl: config.url,
    serviceName: '@openneuro/app',
    serviceVersion: version,
    environment: config.sentry.environment,
  })
} else {
  setupApm({
    serverUrl: config.url,
    serviceName: '@openneuro/app',
    serviceVersion: version,
    environment: config.sentry.environment,
    active: false,
  })
}

gtag.initialize(config.analytics.trackingIds)

Sentry.init({
  dsn: 'https://ba0c58863b3e40a2a412132bfd2711ea@sentry.io/251076',
  release: version,
  environment: config.sentry.environment,
})

// Setup the service worker
if ('serviceWorker' in navigator) {
  //navigator.serviceWorker.register('/sw.js')
} else {
  Sentry.captureMessage('Service worker registration failed.')
}

ReactDOM.render(
  <App>
    <ApolloProvider
      client={createClient(`${config.url}/crn/graphql`, {
        clientVersion: version,
        // @ts-expect-error
        cache: new InMemoryCache().restore(JSON.parse(window.__APOLLO_STATE__)),
      })}>
      <BrowserRouter>
        <Route component={analyticsWrapper(Index)} />
      </BrowserRouter>
    </ApolloProvider>
  </App>,
  document.getElementById('main'),
)
