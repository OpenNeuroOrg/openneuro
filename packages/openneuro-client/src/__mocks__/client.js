import { ApolloClient, InMemoryCache } from '@apollo/client'
import { SchemaLink } from '@apollo/client/link/schema'
import { addMockFunctionsToSchema, makeExecutableSchema } from 'apollo-server'
import { typeDefs } from '@openneuro/server/src/graphql/schema'
import resolvers from '@openneuro/server/src/graphql/resolvers/index.js'

/**
 * Generate sequential dataset ids for tests
 * @param {number} n Starting string value
 */
function* idGenerator(n = 0) {
  while (true) {
    const idStr = String(++n).padStart(6, '0')
    yield `ds${idStr}`
  }
}

export const datasetIds = idGenerator()
export const testDsId = datasetIds.next().value
export const testTime = new Date()

// Workaround to restore this test, it should be refactored with openneuro-client updates
const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

addMockFunctionsToSchema({
  schema,
  mocks: {
    Dataset: (root, { id }) => ({
      id: () => (id ? id : datasetIds.next().value),
      created: testTime,
      modified: testTime,
    }),
    DateTime: () => {
      return testTime
    },
    BigInt: () => {
      return 8589934592
    },
  },
})

export const createClient = uri => {
  const cache = new InMemoryCache()
  const link = new SchemaLink({
    schema,
  })
  // @ts-expect-error
  return new ApolloClient({ uri, link, cache })
}
