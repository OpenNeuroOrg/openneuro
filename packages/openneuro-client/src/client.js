import fetch from 'node-fetch'
import ApolloClient from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createUploadLink } from 'apollo-upload-client/src'
import FormData from 'form-data'
import * as files from './files'
import * as datasets from './datasets'

const cache = new InMemoryCache()

/**
 * Setup a client for working with the OpenNeuro API
 *
 * @param {string} uri
 */
const createClient = uri => {
  const link = createLink(uri)
  return new ApolloClient({ uri, link, cache })
}

const createLink = uri => {
  return createUploadLink({ uri, fetch, serverFormData: FormData })
}

export { files, datasets }
export default createClient
