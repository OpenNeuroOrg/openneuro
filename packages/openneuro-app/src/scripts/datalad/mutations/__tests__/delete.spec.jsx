import React from 'react'
import { render, waitFor, fireEvent } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import { shallow } from 'enzyme'
import DeleteDataset from '../delete.jsx'
import DeleteDir, { DELETE_FILES } from '../delete-dir.jsx'

describe('DeleteDataset mutation', () => {
  it('renders with common props', () => {
    const wrapper = shallow(<DeleteDataset datasetId="ds001" />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('DeleteDir mutation', () => {
  it('renders with common props', () => {
    const wrapper = shallow(
      <DeleteDir
        datasetId="ds002"
        fileTree={{
          files: [],
          directories: [],
          path: '',
        }}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })
  it('fires the correct mutation', async done => {
    const datasetId = 'ds999999'
    const path = 'sub-99'
    const mocks = [
      {
        request: {
          query: DELETE_FILES,
          variables: {
            datasetId,
            files: [{ path: 'sub-99' }],
          },
        },
        newData: jest.fn(() => ({
          data: {},
        })),
      },
    ]
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DeleteDir {...{ datasetId, path }} />
      </MockedProvider>,
    )

    // click "Delete" button
    fireEvent.click(container.querySelector('button.warning'))
    // confirm delete
    fireEvent.click(container.querySelector('button.success'))

    await waitFor(() => expect(mocks[0].newData).toHaveBeenCalled())
    done()
  })
})
