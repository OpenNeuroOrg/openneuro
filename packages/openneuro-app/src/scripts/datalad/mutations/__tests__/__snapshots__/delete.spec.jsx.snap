// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteDataset mutation renders with common props 1`] = `
<Mutation
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "deleteDataset",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "reason",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "reason",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "redirect",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "redirect",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "deleteDataset",
                },
                "selectionSet": undefined,
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "ID",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "reason",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "redirect",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 143,
        "start": 0,
      },
    }
  }
>
  <Component />
</Mutation>
`;

exports[`DeleteDir mutation renders with common props 1`] = `
<Mutation
  awaitRefetchQueries={true}
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "deleteFiles",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "datasetId",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "datasetId",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "files",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "files",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "deleteFiles",
                },
                "selectionSet": undefined,
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "ID",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "datasetId",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "ListType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "DeleteFile",
                    },
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "files",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    }
  }
>
  <Component />
</Mutation>
`;
